src/Orleans.Runtime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -437,35 +436,89 @@ private bool IsValidSilo(SiloAddress silo)
             return !this.membershipSnapshot.ClusterMembership.GetSiloStatus(silo).IsTerminating();
         }
 
-        public SiloAddress CheckIfShouldForward(GrainId grainId, int hopCount, string operationDescription)
+        internal ValueTask<SiloAddress> GetPartitionOwner(GrainId grain, int hopCount, MembershipVersion targetMembershipVersion, string operationDescription)
         {
-            SiloAddress owner = this.membershipSnapshot.CalculateGrainDirectoryPartition(grainId);
-
-            if (owner == null)
+            if (!this.initialStabilizationComplete || targetMembershipVersion != default && this.DirectoryMembershipSnapshot.ClusterMembership.Version < targetMembershipVersion)
             {
-                // There are no available silos for placement, so throw.
-                throw new InvalidOperationException("Grain directory is not operational");
+                return GetPartitionOwnerAsync(grain, targetMembershipVersion, operationDescription);
             }
 
-            if (owner.Equals(MyAddress))
+            return new ValueTask<SiloAddress>(GetPartitionOwnerInternal(grain, operationDescription));
+
+            async ValueTask<SiloAddress> GetPartitionOwnerAsync(GrainId grainId, MembershipVersion membershipVersion, string operation)
             {
-                // if I am the owner, perform the operation locally
-                return null;
+                await this.RefreshMembership(membershipVersion);
+
+                var result = GetPartitionOwnerInternal(grainId, operation);
+
+                // If this silo is supposed to service the request, wait until initial stabilization has completed.
+                if (this.MyAddress.Equals(result) && !this.initialStabilizationComplete)
+                {
+                    var remainingWaitCycles = 80;
+
+                    // Wait for handoff only if there is more than one member in the cluster.
+                    while (this.DirectoryMembershipSnapshot.ActiveMemberCount > 1
+                        && !this.HandoffManager.HasReceivedSplit
+                        && remainingWaitCycles-- > 0)
+                    {
+                        await TimerManager.Delay(TimeSpan.FromMilliseconds(50));
```

Comment:
I wonder if there is a heuristic for this `50 ms` and the `80` cycles? If there is, maybe name constants. Otherwise probably OK as in "shaked something out of stetson".

Reply:
It's fairly arbitrary